src/tequila/ml/interface_torch.py

```diff
@@ -0,0 +1,237 @@
+
+    Parameters
+    ----------
+    objective: Objective:
+        the Objective to be transformed into a torch layer.
+    compile_args: dict:
+        a dictionary of arguments for the tequila compiler; used to render objectives callable
+    input_vars: list:
+        a list of variables; indicates which variables' values will be considered input to the layer.
+
+    Returns
+    -------
+    tuple:
+        the requisite pytorch autograd function, alongside necessary information for higher level classes.
+    """
+
+    comped_objective, compile_args, weight_vars, w_grads, i_grads, first, second \
+        = preamble(objective, compile_args, input_vars)
+    samples = compile_args['samples']
+
+    class _TorchFunction(torch.autograd.Function):
+        """
+        Internal class for the forward and backward passes of calling a tequila objective.
+
+        Notes
+        -----
+
+        Though this class is a private class, some explanations of it's implementation may benefit
+        curious users and later developers.
+
+        Question: why is this class defined within a function?
+        answer: because, since its defined entirely with staticmethods -- about which torch is quite particular --
+        it is impossible to use the attribute 'self' to store information for later use.
+        This means that, to wrap arbitrary tequila Objectives into this class, the class needs to see the in some
+        kind of well contained scope; the function containing this class, get_torch_function, provides that scope.
+        in particular, this scoping is used to associate, with this function, arbitrary tequila objectives,
+        their gradients with respect to weights or inputs (that is, variables specified to be one or the other)
+        and a small dictionary called pattern, which orders the tequila variables w.r.t the order in which a tensor
+        of combined input values and weight values are passed down to the function.
+
+        Though this class doesn't have any proper attributes seperate from those it inherits, we detail
+        the non-torch objects called within the function here:
+
+        For Forward
+        comped_objective: Objective
+            a compiled tequila objective; this function has merely wrapped around it to pass torch Tensors into it.
+
+
+        For Forward and Backward
+        samples: int or None:
+            how many samples the user wants when sampling the Objective or it's gradients.
+
+
+        methods called:
+            tensor_fix:
+                takes a tensor and an (int: Variable) dict and returns a (Variable: float) dict.
+
+        """
+        @staticmethod
+        def forward(ctx, inputs, angles):
+            """
+            forward pass of the function.
+            """
+            ctx.save_for_backward(inputs, angles)
+            call_args = tensor_fix(inputs, angles, first, second)
+            result = comped_objective(variables=call_args, samples=samples)
```

Comment:
dangerous. Better make the comped_objective a member, same for tensor_fix function

Reply:
it is specifically impossible, as far as I know, to do that (I've tried to several times in the past and had to come up with this workaround). There is no ability to access custom "self" attributes in either forward or backward for this function; note that these static methods don't have a self. That's why I have this class as a private class, to begin with. It would be fine to move tensor_fix into this function as a local function, however, and shield it from the user.